javascript_nodejs/5.multiturn-prompts-bot/dialogs/mainDialog/index.js

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+const { ActionTypes, MessageFactory } = require('botbuilder');
+
+const { TextPrompt, NumberPrompt, DialogSet, WaterfallDialog } = require('botbuilder-dialogs');
+
+const DIALOG_STATE_PROPERTY = 'dialogState';
+const USER_NAME_PROP = 'user_name';
+const AGE_PROP = 'user_age';
+const WHO_ARE_YOU = 'who_are_you';
+const HELLO_USER = 'hello_user';
+
+const NAME_PROMPT = 'name_prompt';
+const AGE_PROMPT = 'age_prompt';
+
+class MainDialog {
+    /**
+     * 
+     * @param {Object} conversationState 
+     * @param {Object} userState 
+     */
+    constructor (conversationState, userState) {
+
+        // creates a new state accessor property. see https://aka.ms/about-bot-state-accessors to learn more about the bot state and state accessors 
+        this.conversationState = conversationState;
+        this.userState = userState;
+
+        this.dialogState = this.conversationState.createProperty(DIALOG_STATE_PROPERTY);
+
+        this.userName = this.userState.createProperty(USER_NAME_PROP);
+        this.userAge = this.userState.createProperty(AGE_PROP);
```

Comment:
alternatively we could have an object with name and age and make that object a property?
I'm not sure about the pros and cons. for now I'll copy what you've done

Reply:
@Stevenic any feedback here on whether we should have an object with 2 properties or 2 separate properties?

Reply:
@johnataylor, @benbrown 

Not @stevenic, you've got another Steven instead 😄 

Using an object is more "normal" and for C# and TypeScript what I'd expect because I can be more prescriptive. _"I need user data, and this is all of the user data I expect."_

I'd argue that 99% of all StatePropertyAccessors should probably get and set typed objects, and nothing else. If I have multiple instances of user-related data, I want all of the user data to be accessible at one time. 

The added benefit of using an object is because it's passed by reference in JavaScript, so if a save on the BotState instance is called later on the turn, you don't need to manually call `StatePropertyAccessor.set(turnContext, value);`

Reply:
Fair! Easier to manage than N-number of fields too.

I'll update this example and the validation example as well to use a single accessor to store a user object.

Reply:
OK! I've updated this to use a single accessor.